File: java/SpanishConverter.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

public class SpanishConverter {
    private static String callSoap(int n) throws Exception {
        String soapRequest = "<?xml version=\"1.0\" encoding=\"utf-8\"?>" +
                "<soap:Envelope xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" " +
                "xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" " +
                "xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\">" +
                "<soap:Body>" +
                "<NumberToWords xmlns=\"http://www.dataaccess.com/webservicesserver/\">" +
                "<ubiNum>" + n + "</ubiNum>" +
                "</NumberToWords>" +
                "</soap:Body>" +
                "</soap:Envelope>";

        URL url = new URL("https://www.dataaccess.com/webservicesserver/NumberConversion.wso");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setDoOutput(true);
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "text/xml; charset=utf-8");
        conn.setRequestProperty("SOAPAction", "http://www.dataaccess.com/webservicesserver/NumberToWords");

        conn.getOutputStream().write(soapRequest.getBytes(StandardCharsets.UTF_8));

        InputStream stream = conn.getResponseCode() >= 400 ? conn.getErrorStream() : conn.getInputStream();
        if (stream == null) {
            throw new IOException("Empty SOAP response");
        }

        StringBuilder response = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
        }

        String xml = response.toString();
        String result = extractTag(xml, "NumberToWordsResult");
        if (result.isEmpty()) {
            throw new IOException("SOAP parse error: " + xml);
        }
        return result;
    }

    private static String extractTag(String xml, String tag) {
        String open = "<" + tag + ">";
        String close = "</" + tag + ">";
        int start = xml.indexOf(open);
        if (start == -1) {
            open = "<m:" + tag + ">";
            close = "</m:" + tag + ">";
            start = xml.indexOf(open);
        }
        if (start == -1) {
            return "";
        }
        int end = xml.indexOf(close, start);
        if (end == -1) {
            return "";
        }
        return xml.substring(start + open.length(), end).trim();
    }

    private static String translateText(String text) throws IOException {
        String endpoint = "https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q=" + URLEncoder.encode(text, StandardCharsets.UTF_8);
        URL url = new URL(endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            String json = response.toString();
            int firstQuote = json.indexOf('"');
            int secondQuote = json.indexOf('"', firstQuote + 1);
            if (firstQuote >= 0 && secondQuote > firstQuote) {
                return json.substring(firstQuote + 1, secondQuote);
            }
            return "";
        }
    }

    public static void main(String[] args) throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress(8080), 0);
        server.createContext("/", new HttpHandler() {
            @Override
            public void handle(HttpExchange exchange) throws IOException {
                URI requestURI = exchange.getRequestURI();
                String query = requestURI.getQuery();
                int n = 10;
                if (query != null && query.contains("n=")) {
                    try {
                        String value = query.split("n=")[1].split("&")[0];
                        n = Integer.parseInt(value);
                    } catch (Exception ignored) {
                    }
                }

                String english;
                String spanish;
                try {
                    english = callSoap(n);
                    spanish = translateText(english);
                } catch (Exception ex) {
                    String error = "Error: " + ex.getMessage();
                    exchange.sendResponseHeaders(500, error.getBytes(StandardCharsets.UTF_8).length);
                    try (OutputStream os = exchange.getResponseBody()) {
                        os.write(error.getBytes(StandardCharsets.UTF_8));
                    }
                    return;
                }

                exchange.sendResponseHeaders(200, spanish.getBytes(StandardCharsets.UTF_8).length);
                try (OutputStream os = exchange.getResponseBody()) {
                    os.write(spanish.getBytes(StandardCharsets.UTF_8));
                }
            }
        });
        server.start();
        System.out.println("Servidor escuchando en http://localhost:8080/?n=10");
    }
}

File: go/spanish_converter.go
package main

import (
    "encoding/json"
    "fmt"
    "io"
    "log"
    "net/http"
    "net/url"
    "strconv"
    "strings"
)

func translateText(text string) (string, error) {
    endpoint := "https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q=" + url.QueryEscape(text)
    resp, err := http.Get(endpoint)
    if err != nil {
        return "", err
    }
    defer resp.Body.Close()

    var result any
    if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
        return "", err
    }

    data, ok := result.([]any)
    if !ok || len(data) < 1 {
        return "", fmt.Errorf("unexpected response format")
    }

    first, ok := data[0].([]any)
    if !ok || len(first) < 1 {
        return "", fmt.Errorf("unexpected response format")
    }

    firstEntry, ok := first[0].([]any)
    if !ok || len(firstEntry) < 1 {
        return "", fmt.Errorf("unexpected response format")
    }

    translated, ok := firstEntry[0].(string)
    if !ok {
        return "", fmt.Errorf("unexpected response format")
    }
    return translated, nil
}

func callSoapNumberToWords(n int) (string, error) {
    soapURL := "https://www.dataaccess.com/webservicesserver/NumberConversion.wso"
    soapBody := fmt.Sprintf(`<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>%d</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>`, n)

    req, err := http.NewRequest(http.MethodPost, soapURL, strings.NewReader(soapBody))
    if err != nil {
        return "", err
    }
    req.Header.Set("Content-Type", "text/xml; charset=utf-8")
    req.Header.Set("SOAPAction", "http://www.dataaccess.com/webservicesserver/NumberToWords")

    resp, err := http.DefaultClient.Do(req)
    if err != nil {
        return "", err
    }
    defer resp.Body.Close()

    raw, err := io.ReadAll(resp.Body)
    if err != nil {
        return "", err
    }

    if resp.StatusCode < 200 || resp.StatusCode >= 300 {
        return "", fmt.Errorf("HTTP Status %d: %s", resp.StatusCode, string(raw))
    }

    body := string(raw)
    startTag := "<NumberToWordsResult>"
    endTag := "</NumberToWordsResult>"
    start := strings.Index(body, startTag)
    if start == -1 {
        startTag = "<m:NumberToWordsResult>"
        endTag = "</m:NumberToWordsResult>"
        start = strings.Index(body, startTag)
    }
    if start == -1 {
        return "", fmt.Errorf("SOAP response parse error: missing NumberToWordsResult")
    }
    start += len(startTag)
    end := strings.Index(body[start:], endTag)
    if end == -1 {
        return "", fmt.Errorf("SOAP response parse error: missing NumberToWordsResult closing tag")
    }

    return strings.TrimSpace(body[start : start+end]), nil
}

func handler(w http.ResponseWriter, r *http.Request) {
    n := 10
    if qs := r.URL.Query().Get("n"); qs != "" {
        if parsed, err := strconv.Atoi(qs); err == nil {
            n = parsed
        }
    }

    english, err := callSoapNumberToWords(n)
    if err != nil {
        http.Error(w, err.Error(), http.StatusInternalServerError)
        return
    }

    spanish, err := translateText(english)
    if err != nil {
        http.Error(w, err.Error(), http.StatusInternalServerError)
        return
    }

    fmt.Fprint(w, spanish)
}

func main() {
    http.HandleFunc("/", handler)
    fmt.Println("Servidor iniciado en http://localhost:8080")
    log.Fatal(http.ListenAndServe(":8080", nil))
}
